docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+
+#### ALC Unloading
+During the `AssemblyLoadContext.Unloading` event, the `ClientObject` can remove any reference it has to the `ServerObject`, allowing for the server to unload with the ALC. Since there are no other connections into the ALC other then the connection between client and server, the target ALC should be able to unload everything in itself correctly. Afterwards, the `ProxyObject` could remove its reference to the `ClientObject`, opening it up to be collected by the GC. 
+
+There's no need for the server to dereference the target object, since they are both within the target ALC, and can be removed during the unload process.
+
+There are a few potential places where this isolation isn't perfect, which is discussed in the next point.
+
+#### Call-by-value vs more proxy creation
+Let's say you want to call method `M()` on your `TargetObject`, which returns an object of type `A`, a non-primitive. When it returns the object from the target ALC (where `A` was created) back to the user ALC (where we made the original call for `M()`), we've created an extra reference between the ALCs that we need to deal with. There are a few ways we could try and fix this issue:
+* Don't support non-primitive objects being passed to and from the ALCs
+This isn't recommended since it limits use of the target ALC by quite a lot, but for a quick and dirty version, it could work.
+* For each non-primitive object, generate another proxy that passes back to the user ALC instead of the reference to the normal object. When we unload, we can then remove this proxy as well, keeping the isolation barrier intact.
+This is very similar to how AppDomains handle MarshallByRef objects, and it keeps the functionality the same, but also adds a lot more overhead from creating the proxies during runtime for every non-primitive object.
+* Keep everything call-by-value, by making sure any `ServerObject` returns to the user ALC is the metadata of any object, that gets reconstructed when returning to `ClientObject`
+This option is nice since it works for both inter-process and in-process communication. However, having every object only able to return by value means that we'll be limited to making calls changing objects to a single ALC, and the multiple copies of an object may increase space and cause performance issues.
+
+For my planned implementation, I'll be trying to implement call-by-value, since it keeps the design consistent for all types (primitive and non-primitive) and for in-proc vs cross-proc. However, due to the design for extensibility, it's possible that anyone who wants to change the API can implement any of these options in their own way, by changing the interactions of `ClientObject` and `ServerObject`. If someone wants to limit interactions between plugins and the main context to primitives only, they're welcome to do so with this design.
```

Comment:
I personally don't see how call-by-value makes it easier. Both in-proc and cross-proc solutions must be able to handle both proxied and value types. The difference between call-by-value and call-via-proxy is really just where do we draw the line between the two, but the technology has to support both.
As such it's probably true that whichever version you pick would demonstrate the solution for both anyway.

Reply:
The one thing I was thinking about is how if those proxied types start returning/using other types, do we also need to generate proxies for those? It feels like we'd need to make sure that every proxy would be able to generate additional proxies whenever new objects are used in similar ways, which could cause a lot of complications. Correct me if I'm wrong though.

Reply:
>The one thing I was thinking about is how if those proxied types start returning/using other types, do we also need to generate proxies for those?

Yes in most cases. This is how COM handles this and as you mentioned it "cause[s] a lot of complications".

Reply:
I guess my point is that you will run into trouble either way:
* If you do everything (which is possible) by-value, then you run into the problem of duplication and keeping the state in sync. I actually don't know if there's a solution for the problem I describe below...
* If you do as much as possible by-ref, then you require on the fly proxy generation (at least in theory, in practice this might be avoidable by restricting the problem space). That does lead to problems... as @AaronRobinsonMSFT  points out...

The by-value problem which I don't know a solution for is like this:
* Client calls a server which returns object A - by value (A is declared as class and acts as by-ref object in C#)
* Client gets its own copy of A
* Client calls A.SetName() - since it has its own copy this is a purely client side operation
* Client invokes something else on server which relies on the state of A - now client and server have different view of state of A and might behave very differently.

You could "solve" this by restricting supported scenarios (As in client has to pass A back to the server somehow). But this is not enforecable automatically - it's something the developer must know about. I think that automatic proxy generation while complex might have a much easier job of restricting complex scenarios automatically (by throwing).

Basically I would rather have something which throws if it knows it can lead to issues, then something which lets you get to the bad spot without giving any warnings along the way.